### ISSUE TITLE
BodyParser does not parse consistently.

### ISSUE THREAD

**Issue by joyychang** [IBM-Swift/Kitura#692]
```swift
import Kitura
import KituraNet

router.all(middleware: BodyParser())

router.post("/api/snapshot") { request, response, next in
    guard let body = request.body else {
        try response.status(HTTPStatusCode.badRequest).end()
        Log.error("No body found in request")
        return
    }
    
    guard case let .json(json) = body else {
        try response.status(HTTPStatusCode.badRequest).end()
        Log.error("Body is invalid JSON")
        return
    }

    try response.send(json: json).end()
}
```

I got correct response the first time every time I restart server. Then I get "No body found in request" for all subsequent requests.

**Comment by shmuelk** [comment]
@joyychang Can you provide an example payload that was failing and how it was being sent as well?

**Comment by joyychang** [comment]
Request Headers:
```
POST /api/snapshot HTTP/1.1
Host: localhost:3000
Connection: keep-alive
Content-Length: 277
Accept: application/json, text/plain, */*
Origin: http://localhost:3000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36
Content-Type: application/json;charset=UTF-8
DNT: 1
Referer: http://localhost:3000/
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.8,zh-TW;q=0.6,zh;q=0.4,zh-CN;q=0.2
Cookie: verifyCookie=verified; sandbox-sess=s%3A3qnUJmZscRQM5vMdhotpiYynr4RbAbAa.AyNE43wUvnfN%2BCLIsGQboQD5v1Bm1Gvp%2BW8%2B1BITi4Q; _ga=GA1.1.1602086521.1468940738
```

Request payload:
```
{"id":null,"mode":"advanced","files":[{"name":"HelloWolrd.swift","code":"printf(\"Hello World\")"},{"name":"NewFile2.swift","code":""},{"name":"NewFile3.swift","code":""},{"name":"NewFile4.swift","code":""}],"version":{"name":"DEVELOPMENT-SNAPSHOT-2016-07-25-a","latest":true}}
```

Response Headers first time:
```
HTTP/1.1 200 OK
Content-Type: application/json
Content-Length: 430
Date: Thu, 18 Aug 2016 02:07:23 GMT
Connection: Keep-Alive
Keep-Alive: timeout=60, max=19
```

Response first time:
```json
{
  "files" : [
    {
      "name" : "HelloWolrd.swift",
      "code" : "printf(\"Hello World\")"
    },
    {
      "name" : "NewFile2.swift",
      "code" : ""
    },
    {
      "name" : "NewFile3.swift",
      "code" : ""
    },
    {
      "name" : "NewFile4.swift",
      "code" : ""
    }
  ],
  "mode" : "advanced",
  "version" : {
    "name" : "DEVELOPMENT-SNAPSHOT-2016-07-25-a",
    "latest" : true
  },
  "id" : null
}
```

Response Header subsequent times:
```
HTTP/1.1 400 Bad Request
Date: Thu, 18 Aug 2016 02:07:25 GMT
Content-Length: 0
Connection: Keep-Alive
Keep-Alive: timeout=60, max=18
```

**Comment by youming-lin** [comment]
I tested this using `curl -X POST -H "Content-type: application/json" -d '{"a": "test"}' localhost:3000/api/snapshot` and I get `{"a" : "test" }` back every time.

curl:
```
ylin@youming-mbpr:~/Swift/sandbox-ui$ curl -v -X POST -H "Content-type: application/json" -d '{"a": "test"}' localhost:3000/api/snapshot
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> POST /api/snapshot HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.43.0
> Accept: */*
> Content-type: application/json
> Content-Length: 13
> 
* upload completely sent off: 13 out of 13 bytes
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 18
< Date: Thu, 18 Aug 2016 15:43:30 GMT
< Connection: Keep-Alive
< Keep-Alive: timeout=60, max=19
< 
{
  "a" : "test"
* Connection #0 to host localhost left intact
}
```

Browser (via DHC extension):
<img width="457" alt="screen shot 2016-08-18 at 10 46 16 am" src="https://cloud.githubusercontent.com/assets/19825876/17780490/15884940-6531-11e6-8c87-1f7078cecdd6.png">

**Comment by carlbrown** [comment]
`curl` can take multiple URLs on the same command line, and that seems to be related to the issue.  So if I do:
```
$  curl -X POST -H "Content-type: application/json" -d '{"a": "test"}' http://localhost:3000/api/snapshot  http://localhost:3000/api/snapshot
{
  "a" : "test"
}
$  curl -X POST -H "Content-type: application/json" -d '{"a": "test"}' http://lcalhost:3000/api/snapshot  http://localhost:3000/api/snapshot
{
  "a" : "test"
}
```
Then it works, but if I do:
```
$  curl -X POST -H "Content-type: application/json" -d '{"a": "test"}' http://localhost:3000/api/snapshot  http://localhost:3000/api/snapshot
{
  "a" : "test"
}

**Comment by collinhundley** [comment]
I've had this exact same issue for several months now. In my case, the problem specifically occurs when I send the POST requests using Postman. When the requests originate from other clients (i.e. NSURLSession), `BodyParser` seems to work correctly.

The results differ based on the platform hosting the Kitura server:

1) macOS: The first POST request correctly parses the HTTP body. All following requests produce a `nil` body.

2) Ubuntu: All POST requests are successfully parsed, but the following error is logged each time:

```sh
ERROR: removeIdleSockets() IncomingSocketManager.swift line 161 - epoll_ctl failure. Error code=9. Reason=Bad file descriptor
```

**Comment by shmuelk** [comment]
The above commit fixes the problem. It may take a couple of days to get the fix integrated.

In the meantime you can fix things locally by editing the file:
       Packages/Kitura-net-0.xx.yy/Sources/KituraNet/HTTP/HTTPIncomngMessage.swift

In the function reset() add the following line:

      saveBody = true

Rebuild and the issue should be fixed.

**Comment by joyychang** [comment]
Works for me. Waiting for it to be tagged.

**Comment by ianpartridge** [comment]
Merged. Fix will be in 0.28, which will be tagged in the next few days.
